core/ir/riscv64/codec.h

```diff
@@ -0,0 +1,207 @@
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Rivos, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL RIVOS, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef CODEC_H
+#define CODEC_H 1
+
+#include "decode.h"
+
+/**
+ * @file codec.h
+ * @brief Interface to the RISC-V instruction codec (encoder/decoder).
+ *
+ * Note that encoder does not verify validity of operand values. This is because
+ * currently invalid bit encodings are either reserved for future extensions
+ * (i.e. C.ADDI or C.SLLI) and execute as NOP or HINT instructions (which might
+ * be used as vendor ISA extensions).
+ */
+
+/**
+ * List of ISA extensions handled by the codec.
+ */
+typedef enum {
+    RISCV64_ISA_EXT_RV32A,
+    RISCV64_ISA_EXT_RV32C,
+    RISCV64_ISA_EXT_RV32D,
+    RISCV64_ISA_EXT_RV32F,
+    RISCV64_ISA_EXT_RV32H,
+    RISCV64_ISA_EXT_RV32I,
+    RISCV64_ISA_EXT_RV32M,
+    RISCV64_ISA_EXT_RV32Q,
+    RISCV64_ISA_EXT_RV32ZBA,
+    RISCV64_ISA_EXT_RV32ZBB,
+    RISCV64_ISA_EXT_RV32ZBC,
+    RISCV64_ISA_EXT_RV32ZBS,
+    RISCV64_ISA_EXT_RV64A,
+    RISCV64_ISA_EXT_RV64C,
+    RISCV64_ISA_EXT_RV64D,
+    RISCV64_ISA_EXT_RV64F,
+    RISCV64_ISA_EXT_RV64H,
+    RISCV64_ISA_EXT_RV64I,
+    RISCV64_ISA_EXT_RV64M,
+    RISCV64_ISA_EXT_RV64Q,
+    RISCV64_ISA_EXT_RV64ZBA,
+    RISCV64_ISA_EXT_RV64ZBB,
+    RISCV64_ISA_EXT_RVC,
+    RISCV64_ISA_EXT_SVINVAL,
+    RISCV64_ISA_EXT_SYSTEM,
+    RISCV64_ISA_EXT_ZICBOM,
+    RISCV64_ISA_EXT_ZICBOP,
+    RISCV64_ISA_EXT_ZICBOZ,
+    RISCV64_ISA_EXT_ZICSR,
+    RISCV64_ISA_EXT_ZIFENCEI,
+    RISCV64_ISA_EXT_CNT, /* Keep this last */
+} riscv64_isa_ext_t;
+
+/**
+ * List of instruction formats handled by the codec.
+ *
+ * Note that enum names have to match ones defined in the codec.py.
+ */
+typedef enum {
+    /* Uncompressed instruction formats */
+    RISCV64_FMT_R = 0,
+    RISCV64_FMT_R4,
```

Comment:
I needed to refer `riscv64_inst_fmt_t` while reviewing #5668. Quick comment: are these defined by the RISC-V ISA? Can you link to the relevant ISA documentation that describes what they mean?

Reply:
Sec 2.2, 2.3, etc. here seem to list some templates with letter names: https://riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf
I assume they match these?

Reply:
Exactly. Though I could add ASCII-art comments describing the layout to limit the need to jump to the specification document.

Reply:
Done